File: DisMethod/MoveWall/Fortran/Math.f90
INCLUDE "mkl_dfti.f90"
    
MODULE MATH
    USE MKL_DFTI
    IMPLICIT NONE
    
    INTERFACE TDMA
        MODULE PROCEDURE TDMA_REAL
        MODULE PROCEDURE TDMA_COMPLEX
    END INTERFACE
    
    !SUPPLIMENTAL VECTOR
    REAL, SAVE, ALLOCATABLE :: CC(:), BB(:), DD(:)
    COMPLEX, SAVE, ALLOCATABLE :: C_CC(:), C_BB(:), C_DD(:)
    REAL, SAVE, ALLOCATABLE :: U(:), V(:)
    INTEGER, SAVE :: LENGTH = -1
    LOGICAL, SAVE :: ALL_ALLOCATED
    
    !FFT/IFFT SUPPLIMENTAL ARRAY
    COMPLEX, SAVE, ALLOCATABLE :: PF_1D(:)
    COMPLEX, SAVE, ALLOCATABLE :: SF_1D(:)
    COMPLEX, SAVE, ALLOCATABLE :: SF2(:, :, :)
    COMPLEX, SAVE, ALLOCATABLE :: SHIFT(:, :, :)
    INTEGER, SAVE :: FFT_DIM(3) = -1, FFT_LEN = -1
    TYPE(DFTI_DESCRIPTOR), SAVE, POINTER :: HANDLER
    INTEGER, SAVE :: STATUS
    LOGICAL, SAVE :: FFT_ALLOCATED = .FALSE.
    
    PRIVATE BB, CC, DD, U, V, LENGTH, ALL_ALLOCATED
    PRIVATE PF_1D, SF_1D, FFT_ALLOCATED
    
    CONTAINS
    
    SUBROUTINE INITVEC(N)
        IMPLICIT NONE
        INTEGER, INTENT(IN) :: N
        
        IF(ALL_ALLOCATED) THEN
            IF(N <= LENGTH) THEN
                RETURN
            END IF
            DEALLOCATE(BB, CC, DD, U, V)
            DEALLOCATE(C_BB, C_CC, C_DD)
            ALLOCATE(BB(N))
            ALLOCATE(CC(N))
            ALLOCATE(DD(N))
            ALLOCATE(C_BB(N))
            ALLOCATE(C_CC(N))
            ALLOCATE(C_DD(N))
            ALLOCATE(U(N))
            ALLOCATE(V(N))
            LENGTH = N
        ELSE
            ALLOCATE(BB(N))
            ALLOCATE(CC(N))
            ALLOCATE(DD(N))
            ALLOCATE(C_BB(N))
            ALLOCATE(C_CC(N))
            ALLOCATE(C_DD(N))
            ALLOCATE(U(N))
            ALLOCATE(V(N))
            LENGTH = N
            ALL_ALLOCATED = .TRUE.
        END IF
    END SUBROUTINE INITVEC
    
    SUBROUTINE DELETEVEC()
        IMPLICIT NONE
        IF(ALL_ALLOCATED) THEN
            DEALLOCATE(BB, CC, DD, U, V)
            DEALLOCATE(C_BB, C_CC, C_DD)
            LENGTH = 0
        END IF
    END SUBROUTINE DELETEVEC
            
    SUBROUTINE TDMA_REAL(A, B, C, X, R)
        IMPLICIT NONE
        REAL, INTENT(IN) :: A(:), B(:), C(:), R(:)
        REAL, INTENT(OUT) :: X(:)
        
        INTEGER I, N
        
        N = SIZE(X)
        CALL INITVEC(N)
        
        CC(1) = C(1) / B(1)
        DO I = 2, N - 1
            CC(I) = C(I) / (B(I) - A(I) * CC(I - 1))
        END DO
        
        BB = 1
        
        DD(1) = R(1) / B(1)
        DO I = 2, N
            DD(I) = (R(I) - A(I) * DD(I - 1)) / (B(I) - A(I) * CC(I - 1))
        END DO
        
        X(N) = DD(N)
        DO I = N - 1, 1, -1
            X(I) = DD(I) - CC(I) * X(I + 1)
        END DO
    END SUBROUTINE TDMA_REAL
    
    SUBROUTINE TDMA_COMPLEX(A, B, C, X, R)
        IMPLICIT NONE
        COMPLEX, INTENT(IN) :: A(:), B(:), C(:), R(:)
        COMPLEX, INTENT(OUT) :: X(:)
        
        REAL, ALLOCATABLE :: D(:)
        REAL SUMR
        INTEGER I, N
        
        N = SIZE(X)
        !ALLOCATE(D(N))
        !D = REAL(A)
        !D = REAL(B)
        !D = REAL(C)
        !D = REAL(R)
        !
        !D(1) = REAL(A(2) + B(1))
        !DO I = 2, N-1
        !    D(I) = A(I + 1) + B(I) + C(I - 1)
        !END DO
        !D(N) = REAL(B(N) + C(N-1))
        !
        !SUMR = SUM(REAL(R))
        
        CALL INITVEC(N)
        
        C_CC(1) = C(1) / B(1)
        DO I = 2, N - 1
            C_CC(I) = C(I) / (B(I) - A(I) * C_CC(I - 1))
        END DO
        
        C_BB = 1
        
        C_DD(1) = R(1) / B(1)
        DO I = 2, N
            C_DD(I) = (R(I) - A(I) * C_DD(I - 1)) / (B(I) - A(I) * C_CC(I - 1))
        END DO
        
        X(N) = C_DD(N)
        DO I = N - 1, 1, -1
            X(I) = C_DD(I) - C_CC(I) * X(I + 1)
        END DO
    END SUBROUTINE TDMA_COMPLEX
    
    SUBROUTINE CTDMA(A, B, C, X, R)
        IMPLICIT NONE
        REAL, INTENT(IN) :: A(:), B(:), C(:), R(:)
        REAL, INTENT(OUT) :: X(:)
        
        REAL XN
        INTEGER I, N
        
        N = SIZE(X)
        CALL INITVEC(N)
        
        CC(1) = C(1) / B(1)
        DO I = 2, N - 2
            CC(I) = C(I) / (B(I) - A(I) * CC(I - 1))
        END DO
        
        BB = 1
        
        DD(1) = -A(1) / B(1)
        DO I = 2, N - 2
            DD(I) = -A(I) * DD(I - 1) / (B(I) - A(I) * CC(I - 1))
        END DO
        DD(N - 1) = (-C(N - 1) - A(I) * DD(I - 1)) / (B(I) - A(I) * CC(I - 1))
        
        U(N - 1) = DD(N - 1)
        DO I = N - 2, 1, -1
            U(I) = DD(I) - CC(I) * U(I + 1)
        END DO
        
        DD(1) = R(1) / B(1)
        DO I = 2, N - 1
            DD(I) = (R(I) - A(I) * DD(I - 1)) / (B(I) - A(I) * CC(I - 1))
        END DO
        
        V(N - 1) = DD(N - 1)
        DO I = N - 2, 1, -1
            V(I) = DD(I) - CC(I) * V(I + 1)
        END DO
        
        XN = (R(N) - C(N) * V(1) - A(N) * V(N - 1)) / &
             (C(N) * U(1) + A(N) * U(N - 1) + B(N))
        
        X = U * XN + V
        X(N) = XN
        
    END SUBROUTINE CTDMA
    
    SUBROUTINE FFT_ALLOC(L)
        IMPLICIT NONE
        INTEGER, INTENT(IN) :: L(3)
        
        IF(FFT_ALLOCATED) THEN
            IF(SIZE(PF_1D) < L(1) * L(3)) THEN
                DEALLOCATE(PF_1D, SF_1D)
                ALLOCATE(PF_1D(L(1) * L(3)))
                ALLOCATE(SF_1D(L(1) * L(3)))
            END IF
            IF(FFT_DIM(1) < L(1) .OR. FFT_DIM(2) < L(2) .OR. FFT_DIM(3) < L(3)) THEN
                DEALLOCATE(SHIFT)
                ALLOCATE(SHIFT(L(1), L(2), L(3)))
                DEALLOCATE(SF2)
                ALLOCATE(SF2(L(1), L(2), L(3)))
            END IF
            
            IF(FFT_DIM(1) .NE. L(1) .OR. FFT_DIM(3) .NE. L(3)) THEN
                STATUS = DFTIFREEDESCRIPTOR(HANDLER)
                STATUS = DFTICREATEDESCRIPTOR(HANDLER, DFTI_DOUBLE, DFTI_COMPLEX, 2, (/L(1), L(3)/))
                STATUS = DFTISETVALUE(HANDLER, DFTI_PLACEMENT, DFTI_NOT_INPLACE)
                STATUS = DFTICOMMITDESCRIPTOR(HANDLER)
            END IF
        ELSE
            ALLOCATE(PF_1D(L(1) * L(3)))
            ALLOCATE(SF_1D(L(1) * L(3)))
            ALLOCATE(SHIFT(L(1), L(2), L(3)))
            ALLOCATE(SF2(L(1), L(2), L(3)))
            STATUS = DFTICREATEDESCRIPTOR(HANDLER, DFTI_DOUBLE, DFTI_COMPLEX, 2, (/L(1), L(3)/))
            STATUS = DFTISETVALUE(HANDLER, DFTI_PLACEMENT, DFTI_NOT_INPLACE)
            STATUS = DFTICOMMITDESCRIPTOR(HANDLER)
            FFT_ALLOCATED = .TRUE.
        END IF
        
        FFT_DIM(1) = L(1)
        FFT_DIM(2) = L(2)
        FFT_DIM(3) = L(3)
        FFT_LEN = L(1) * L(3)
    END SUBROUTINE FFT_ALLOC
    
    SUBROUTINE FFT_DEALLOC()
        IMPLICIT NONE
        
        IF(FFT_ALLOCATED) THEN
            DEALLOCATE(PF_1D, SF_1D)
            DEALLOCATE(SHIFT)
            DEALLOCATE(SF2)
            STATUS = DFTIFREEDESCRIPTOR(HANDLER)
            FFT_ALLOCATED = .FALSE.
        END IF
        
        FFT_DIM = -1
    END SUBROUTINE FFT_DEALLOC
    
    SUBROUTINE FFT(PF, SF)
        IMPLICIT NONE
        REAL(8), INTENT(IN) :: PF(:, :, :)
        COMPLEX(8), INTENT(OUT) :: SF(:, :, :)
        INTEGER J
        
        CALL FFT_ALLOC(SHAPE(PF))
        DO J = 1, SIZE(PF, 2)
            PF_1D(1 : FFT_LEN) = RESHAPE(PF(:, J, :), (/FFT_LEN/))
            STATUS = DFTICOMPUTEFORWARD(HANDLER, PF_1D(1 : FFT_LEN), SF_1D(1 : FFT_LEN))
            SF(:, J, :) = RESHAPE(SF_1D, (/SIZE(SF, 1), SIZE(SF, 3)/))
        END DO
        CALL FFTSHIFT(SF)
    END SUBROUTINE FFT
    
    SUBROUTINE IFFT(SF, PF)
        IMPLICIT NONE
        COMPLEX(8), INTENT(IN) :: SF(:, :, :)
        REAL(8), INTENT(OUT) :: PF(:, :, :)
        INTEGER J
        
        CALL FFT_ALLOC(SHAPE(PF))
        SF2(1:FFT_DIM(1), 1:FFT_DIM(2), 1:FFT_DIM(3)) = SF
        CALL IFFTSHIFT(SF2(1:FFT_DIM(1), 1:FFT_DIM(2), 1:FFT_DIM(3)))
        DO J = 1, SIZE(PF, 2)
            SF_1D(1 : FFT_LEN) = RESHAPE(SF2(1:FFT_DIM(1), J, 1:FFT_DIM(3)), (/FFT_LEN/))
            STATUS = DFTICOMPUTEBACKWARD(HANDLER, SF_1D(1 : FFT_LEN), PF_1D(1 : FFT_LEN))
            PF(:, J, :) = REAL(RESHAPE(PF_1D / SIZE(PF, 1) / SIZE(PF, 3), (/SIZE(PF, 1), SIZE(PF, 3)/)))
        END DO
        
    END SUBROUTINE IFFT
    
    SUBROUTINE FFTSHIFT(SF)
        IMPLICIT NONE
        COMPLEX(8), INTENT(INOUT) :: SF(:, :, :)
        
        INTEGER :: N1H, N3H, N1, N2, N3
        
        N1 = SIZE(SF, 1)
        N2 = SIZE(SF, 2)
        N3 = SIZE(SF, 3)
        N1H = SIZE(SF, 1) / 2
        N3H = SIZE(SF, 3) / 2
        
        SHIFT(1 : N1H, 1:N2, :) = SF(N1 - N1H + 1 : N1, :, :)
        SHIFT(N1H + 1 : N1, 1:N2, :) = SF(1 : N1 - N1H, :, :)
        SF(:, :, 1 : N3H) = SHIFT(:, 1:N2, N3 - N3H + 1 : N3)
        SF(:, :, N3H + 1 : N3) = SHIFT(:, 1:N2, 1 : N3 - N3H)
    END SUBROUTINE FFTSHIFT
    
    SUBROUTINE IFFTSHIFT(SF)
        IMPLICIT NONE
        COMPLEX(8), INTENT(INOUT) :: SF(:, :, :)
        
        INTEGER :: N1H, N3H, N1, N2, N3
        
        N1 = SIZE(SF, 1)
        N2 = SIZE(SF, 2)
        N3 = SIZE(SF, 3)
        N1H = SIZE(SF, 1) / 2
        N3H = SIZE(SF, 3) / 2
        
        SHIFT(:, 1:N2, 1 : N3 - N3H) = SF(:, :, N3H + 1 : N3)
        SHIFT(:, 1:N2, N3 - N3H + 1 : N3) = SF(:, :, 1 : N3H)
        SF(1 : N1 - N1H, :, :) = SHIFT(N1H + 1 : N1, 1:N2, :)
        SF(N1 - N1H + 1 : N1, :, :) = SHIFT(1 : N1H, 1:N2, :)
    END SUBROUTINE IFFTSHIFT
    
END MODULE MATH

File: DisMethod/SteadyWall/C++/Math.cpp
#include "Math.h"
#include "Datatype.h"
#include "mkl_dfti.h"
#include <iostream>
#include <memory>

//supplimental vector
double *cc, *bb, *dd;
double *cc_im, *bb_im, *dd_im;
double *uu, *vv;
int length = -1;
bool all_allocated = false;

//fft/ifft supplimental array
double *pf_1d_re, *pf_1d_im;
double *sf_1d_re, *sf_1d_im;
Field3d<double> sf2_re, sf2_im;
Field3d<double> shift_re, shift_im;
int fft_dim[4], fft_len = -1;
DFTI_DESCRIPTOR_HANDLE handler;
bool fft_allocated = false;

void initvec(int n)
{
	if (all_allocated)
	{
		if (n <= length) goto init;
		delete[] bb, cc, dd, uu, vv;
		delete[] bb_im, cc_im, dd_im;
		bb = new double[n + 1];
		cc = new double[n + 1];
		dd = new double[n + 1];
		bb_im = new double[n + 1];
		cc_im = new double[n + 1];
		dd_im = new double[n + 1];
		uu = new double[n + 1];
		vv = new double[n + 1];
		length = n;
	}
	else
	{
		bb = new double[n + 1];
		cc = new double[n + 1];
		dd = new double[n + 1];
		bb_im = new double[n + 1];
		cc_im = new double[n + 1];
		dd_im = new double[n + 1];
		uu = new double[n + 1];
		vv = new double[n + 1];
		length = n;
		all_allocated = true;
	}
init:
	memset(bb, 0, sizeof(double) * (n + 1));
	memset(cc, 0, sizeof(double) * (n + 1));
	memset(dd, 0, sizeof(double) * (n + 1));
	memset(bb_im, 0, sizeof(double) * (n + 1));
	memset(cc_im, 0, sizeof(double) * (n + 1));
	memset(dd_im, 0, sizeof(double) * (n + 1));
	memset(uu, 0, sizeof(double) * (n + 1));
	memset(vv, 0, sizeof(double) * (n + 1));
}

void deletevec()
{
	if (all_allocated)
	{
		delete[] bb, cc, dd, uu, vv;
		delete[] bb_im, cc_im, dd_im;
		length = 0;
	}
}

void tdma(double *a, double * b, double * c, double *x, double *r, int gap1, int gap2, int n)
{
	//a, b, c start from index-1
	//x, r start from index-0 with gap1 & gap2
	initvec(n);

	cc[1] = c[1] / b[1];
	for (size_t i = 2; i < n; i++)
	{
		cc[i] = c[i] / (b[i] - a[i] * cc[i - 1]);
	}

	dd[1] = r[0] / b[1];
	for (size_t i = 2; i < n+1; i++)
	{
		dd[i] = (r[(i - 1) * gap2] - a[i] * dd[i - 1]) / (b[i] - a[i] * cc[i - 1]);
	}

	x[(n - 1) * gap1] = dd[n];
	for (size_t i = n - 1; i > 0; i--)
	{
		x[(i - 1) * gap1] = dd[i] - cc[i] * x[i * gap1];
	}
}

void tdma(double *a, double* b, double *c, double *x_re, double *x_im, double *r_re, double *r_im, int gap1, int gap2, int n)
{
	//initvec(n);

	////suppose a, b, c are real
	//cc[1] = c[1] / b[1];
	//for (size_t i = 2; i < n; i++)
	//{
	//	cc[i] = c[i] / (b[i] - a[i] * cc[i - 1]);
	//}

	//dd[1] = r_re[0] / b[1];
	//dd_im[1] = r_im[0] / b[1];
	//for (size_t i = 2; i < n+1; i++)
	//{
	//	dd[i] = (r_re[(i - 1) * gap2] - a[i] * cc[i - 1]) / (b[i] - a[i] * cc[i - 1]);
	//	dd_im[i] = (r_im[(i - 1) * gap2] - a[i] * cc[i - 1]) / (b[i] - a[i] * cc[i - 1]);
	//}

	//x_re[(n - 1) * gap1] = dd[n];
	//x_im[(n - 1) * gap1] = dd_im[n];
	//for (size_t i = n - 1; i > 0; i--)
	//{
	//	x_re[(i - 1) * gap1] = dd[i] - cc[i] * x_re[i * gap1];
	//	x_im[(i - 1) * gap1] = dd_im[i] - cc[i] * x_im[i * gap1];
	//}
	tdma(a, b, c, x_re, r_re, gap1, gap2, n);
	tdma(a, b, c, x_im, r_im, gap1, gap2, n);
}

void ctdma(double *a, double *b, double *c, double *x, double *r, int gap1, int gap2, int n)
{
	initvec(n);

	cc[1] = c[1] / b[1];
	for (size_t i = 2; i < n - 1; i++)
		cc[i] = c[i] / (b[i] - a[i] * cc[i - 1]);

	dd[1] = -a[1] / b[1];
	for (size_t i = 2; i < n - 1; i++)
		dd[i] = -a[i] * dd[i - 1] / (b[i] - a[i] * cc[i - 1]);
	dd[n - 1] = (-c[n - 1] - a[n - 1] * dd[n - 2]) / (b[n - 1] - a[n - 1] * cc[n - 2]);

	uu[n - 1] = dd[n - 1];
	for (size_t i = n - 2; i > 0; i--)
		uu[i] = dd[i] - cc[i] * uu[i + 1];

	dd[1] = r[0] / b[1];
	for (size_t i = 2; i < n; i++)
		dd[i] = (r[(i - 1) * gap2] - a[i] * dd[i - 1]) / (b[i] - a[i] * cc[i - 1]);

	vv[n - 1] = dd[n - 1];
	for (size_t i = n - 2; i > 0; i--)
		vv[i] = dd[i] - cc[i] * vv[i + 1];

	double xn = (r[(n - 1) * gap2] - c[n] * vv[1] - a[n] * vv[n - 1]) /
		(c[n] * uu[1] + a[n] * uu[n - 1] + b[n]);

	for (size_t i = 0; i < n - 1; i++)
		x[i * gap1] = uu[i + 1] * xn + vv[i + 1];
	x[(n - 1) * gap1] = xn;
}

void fft_init(int l1, int l2, int l3)
{
	if (!fft_allocated)
	{
		pf_1d_re = new double[l1 * l3];
		pf_1d_im = new double[l1 * l3];
		sf_1d_re = new double[l1 * l3];
		sf_1d_im = new double[l1 * l3];
		shift_im.reAlloc(l1, l2, l3);
		shift_re.reAlloc(l1, l2, l3);
		sf2_im.reAlloc(l1, l2, l3);
		sf2_re.reAlloc(l1, l2, l3);
		MKL_LONG n[2] = { l1, l3 };
		DftiCreateDescriptor(&handler, DFTI_DOUBLE, DFTI_COMPLEX, 2, n);
		DftiSetValue(handler, DFTI_COMPLEX_STORAGE, DFTI_REAL_REAL);
		DftiSetValue(handler, DFTI_PLACEMENT, DFTI_NOT_INPLACE);
		DftiCommitDescriptor(handler);
		fft_allocated = true;

		fft_dim[1] = l1;
		fft_dim[2] = l2;
		fft_dim[3] = l3;
		fft_len = l1 * l3;
		memset(pf_1d_im, 0, sizeof(double) * l1 * l3);
		memset(pf_1d_re, 0, sizeof(double) * l1 * l3);
		memset(sf_1d_im, 0, sizeof(double) * l1 * l3);
		memset(sf_1d_re, 0, sizeof(double) * l1 * l3);
	}
}

void fft_del()
{
	if (fft_allocated)
	{
		delete[] pf_1d_im, pf_1d_re, sf_1d_im, sf_1d_re;
		DftiFreeDescriptor(&handler);
		fft_allocated = false;
		fft_dim[1] = -1;
		fft_dim[2] = -1;
		fft_dim[3] = -1;
	}
}

void fft(Field3d<double>& pf, Field3d<double>& sf_re, Field3d<double>& sf_im)
{
	fft_init(pf.dim1(), pf.dim2(), pf.dim3());
	for (size_t j = 1; j < pf.dim2() + 1; j++)
	{
		int iter = 0;
		for (size_t i = 1; i <= pf.dim1(); i++)
			for (size_t k = 1; k <= pf.dim3(); k++)
				pf_1d_re[iter++] = pf(i, j, k);
		DftiComputeForward(handler, pf_1d_re, pf_1d_im, sf_1d_re, sf_1d_im);
		int iter2 = 0;
		iter = 0;
		for (size_t i = 1; i <= sf_re.dim1(); i++)
			for (size_t k = 1; k <= sf_re.dim3(); k++)
			{
				sf_re(i, j, k) = sf_1d_re[iter++];
				sf_im(i, j, k) = sf_1d_im[iter2++];
			}
	}
	fftshift(sf_re, sf_im);
}

void ifft(Field3d<double>& sf_re, Field3d<double>& sf_im, Field3d<double>& pf)
{
	fft_init(pf.dim1(), pf.dim2(), pf.dim3());
	sf2_im = sf_im;
	sf2_re = sf_re;
	ifftshift(sf2_re, sf2_im);
	int iter = 0;
	int iter2 = 0;
	for (size_t j = 1; j <= pf.dim2(); j++)
	{
		iter = iter2 = 0;
		for (size_t i = 1; i <= pf.dim1(); i++)
			for (size_t k = 1; k <= pf.dim3(); k++)
			{
				sf_1d_re[iter++] = sf2_re(i, j, k);
				sf_1d_im[iter2++] = sf2_im(i, j, k);
			}
		DftiComputeBackward(handler, sf_1d_re, sf_1d_im, pf_1d_re, pf_1d_im);
		iter = iter2 = 0;
		for (size_t i = 1; i <= pf.dim1(); i++)
			for (size_t k = 1; k <= pf.dim3(); k++)
			{
				pf(i, j, k) = pf_1d_re[iter++];
			}
	}
	pf /= (pf.dim1() * pf.dim3());
}

void fftshift(Field3d<double>& sf_re, Field3d<double>& sf_im)
{
	int n1 = sf_re.dim1();
	int n2 = sf_re.dim2();
	int n3 = sf_re.dim3();
	int n1h = n1 / 2;
	int n3h = n3 / 3;

	for (size_t i = 1; i <= n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				shift_re(i, j, k) = sf_re(n1 - n1h + i, j, k);

	for (size_t i = 1; i <= n1 - n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				shift_re(n1h + i, j, k) = sf_re(i, j, k);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3h; k++)
				sf_re(i, j, k) = shift_re(i, j, n3 - n3h + k);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3 - n3h; k++)
				sf_re(i, j, k + n3h) = shift_re(i, j, k);

	for (size_t i = 1; i <= n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				shift_im(i, j, k) = sf_im(n1 - n1h + i, j, k);

	for (size_t i = 1; i <= n1 - n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				shift_im(n1h + i, j, k) = sf_im(i, j, k);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3h; k++)
				sf_im(i, j, k) = shift_im(i, j, n3 - n3h + k);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3 - n3h; k++)
				sf_im(i, j, k + n3h) = shift_im(i, j, k);

}

void ifftshift(Field3d<double>& sf_re, Field3d<double>& sf_im)
{
	int n1 = sf_re.dim1();
	int n2 = sf_re.dim2();
	int n3 = sf_re.dim3();
	int n1h = n1 / 2;
	int n3h = n3 / 3;

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3 - n3h; k++)
				shift_im(i, j, k) = sf_im(i, j, k + n3h);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3h; k++)
				shift_im(i, j, k + n3 - n3h) = sf_im(i, j, k);

	for (size_t i = 1; i <= n1 - n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				sf_im(i, j, k) = shift_im(i + n1h, j, k);

	for (size_t i = 1; i <= n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				sf_im(i + n1 - n1h, j, k) = shift_im(i, j, k);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3 - n3h; k++)
				shift_re(i, j, k) = sf_re(i, j, k + n3h);

	for (size_t i = 1; i <= n1; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3h; k++)
				shift_re(i, j, k + n3 - n3h) = sf_re(i, j, k);

	for (size_t i = 1; i <= n1 - n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				sf_re(i, j, k) = shift_re(i + n1h, j, k);

	for (size_t i = 1; i <= n1h; i++)
		for (size_t j = 1; j <= n2; j++)
			for (size_t k = 1; k <= n3; k++)
				sf_re(i + n1 - n1h, j, k) = shift_re(i, j, k);

}